### ISSUE TITLE
how to overwrite $trafo in paradox 1.0.1

### ISSUE THREAD

**Issue by Luigi-Annic** [mlr-org/paradox#418]
First of all, thank you for this wonderful package.

I am not an expert of this subject, but in the past I used the previous version (0.11.1), where I had a chunk of code like this

```
search_space <- ps(
    dropout = p_dbl(lower = 0, upper = 0.1),
    weight_decay = p_dbl(lower = 0, upper = 0.01),
    learning_rate = p_dbl(lower = 0, upper = 0.001),
    nodes = p_int(lower = 64, upper = 256),
    k = p_int(lower = 1, upper = 4)
  )
  
  search_space$trafo <- function(x, param_set) {
    x$num_nodes = rep(x$nodes, x$k)
    x$nodes = x$k = NULL
    return(x)
  }
```

which was used to overwrite and define my own function for `trafo`.

I know that since the update of this package,  these objects are less mutable, and I have read this discussion (https://github.com/mlr-org/paradox/issues/246) about the matter.

However, I wanted to know if it is possible to get the same result as in the previous code with the updated package; I have seen that there is the possibility to add the `.extra_trafo` argument but I am not sure if this is the way to proceed, and I have not found tutorials or guides about this (example below).

```
  search_space <- ps(
    dropout = p_dbl(lower = 0, upper = 0.1),
    weight_decay = p_dbl(lower = 0, upper = 0.01),
    learning_rate = p_dbl(lower = 0, upper = 0.001),
    nodes = p_int(lower = 64, upper = 256),
    k = p_int(lower = 1, upper = 4),
    .extra_trafo = function(x, param_set) {
      x$num_nodes = rep(x$nodes, x$k)
      x$nodes = x$k = NULL
      return(x)
    }
  )
 ``` 

Thanks,
Luigi

**Comment by mb706** [comment]
`.extra_trafo` is indeed what you use in place of the old `$trafo`. You can also set the `search_space$extra_trafo` field after construction, but the way you do it is the intended way. You can now omit the `param_set` argument if you want, since it is not used by your function.

In cases where the object that you are tuning has a `param_set` already, e.g.
```r
some_objects_param_set <- ps(
  dropout = p_dbl(0, 1),
  weight_decay = p_dbl(0, 1),
  learning_rate = p_dbl(0),
  nodes = p_uty(),
  something_else = p_fct(c("a", "b", "c"))
)
```
you can also get your search space using the `TuneToken` / `to_tune()` machinery. This allows you to specify search spaces for the individual params, where your `.extra_trafo()` only needs to construct the individual component:
```r
search_space <- some_objects_param_set$search_space(list(
  dropout = to_tune(0, 0.1),
  weight_decay = to_tune(0, 0.01),
  learning_rate = to_tune(0, 0.001),
  nodes = to_tune(ps(
    nodes = p_int(64, 256),
    k = p_int(1, 4),
    .extra_trafo = function(x) list(nodes = rep(x$nodes, x$k))
  ))
))
```

You can use `generate_design_random(search_space, n)$transpose()` or `generate_design_grid(search_space, x)$transpose()` to look at what the values that are generated using these trafos generally look like.

As an unrelated comment: Depending on what you are doing with your search space, you may want to use logarithmic scales for some of these hyperparameters (weight decay, learning rate, number of nodes). You can do this with `logscale = TRUE`, both in `p_int` / `p_dbl`, as well as in `to_tune()`. This also uses the "trafo": sampling happens on [log(`lower`), log(`upper`)], which is then transformed using `exp()`, with some special care for integers. If you do this, the lower bounds can not be 0, so you'd need to put some thought into a sensible lower bound.

**Comment by Luigi-Annic** [comment]
Thank you for your help and for your comment!
